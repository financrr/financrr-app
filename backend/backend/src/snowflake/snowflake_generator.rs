use crate::snowflake::snowflake_error::SnowflakeGeneratorError;
use crate::time::timestamp::get_epoch_millis;
use std::sync::atomic::{AtomicU64, Ordering};

pub(crate) const SNOWFLAKE_EPOCH: i64 = 1_705_247_483_000;

pub(crate) const NODE_ID_BITS: u8 = 10;
pub(crate) const SEQUENCE_BITS: u8 = 12;

pub(crate) const TIMESTAMP_SHIFT: u8 = NODE_ID_BITS + SEQUENCE_BITS;

pub(crate) const MAX_NODE_ID: u64 = (1 << NODE_ID_BITS) - 1;
pub(crate) const MAX_SEQUENCE: u64 = (1 << SEQUENCE_BITS) - 1;

#[derive(Debug)]
pub(crate) struct SnowflakeGenerator {
    node_id: u64,
    last_timestamp: AtomicU64,
    sequence: AtomicU64,
}

impl SnowflakeGenerator {
    pub(crate) fn new(node_id: u64) -> Self {
        Self {
            node_id,
            last_timestamp: AtomicU64::new(0),
            sequence: AtomicU64::new(0),
        }
    }

    pub(crate) fn next_id(&self) -> Result<i64, SnowflakeGeneratorError> {
        let mut current_timestamp = self.timestamp();
        let last_timestamp = self.last_timestamp.load(Ordering::SeqCst);

        if current_timestamp < last_timestamp {
            return Err(SnowflakeGeneratorError::InvalidSystemClock);
        }

        let mut sequence = self.sequence.load(Ordering::SeqCst);

        if current_timestamp == last_timestamp {
            sequence = (sequence + 1) & MAX_SEQUENCE;
            if sequence == 0 {
                current_timestamp = self.wait_for_next_millis(current_timestamp, last_timestamp)?;
            }
        } else {
            sequence = 0;
        }

        self.last_timestamp.store(current_timestamp, Ordering::SeqCst);
        self.sequence.store(sequence, Ordering::SeqCst);

        Ok(((current_timestamp << (NODE_ID_BITS + SEQUENCE_BITS)) | (self.node_id << SEQUENCE_BITS) | sequence) as i64)
    }

    fn timestamp(&self) -> u64 {
        (get_epoch_millis() - (SNOWFLAKE_EPOCH as u128)) as u64
    }

    fn wait_for_next_millis(
        &self,
        mut current_timestamp: u64,
        last_timestamp: u64,
    ) -> Result<u64, SnowflakeGeneratorError> {
        while current_timestamp == last_timestamp {
            current_timestamp = self.timestamp();
        }

        Ok(current_timestamp)
    }
}
