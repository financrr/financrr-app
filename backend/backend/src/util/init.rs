use std::fmt::Debug;
use std::process;

use tracing::error;

pub(crate) fn expect_or_exit<T, E: Debug>(result: Result<T, E>, message: &str) -> T {
    match result {
        Ok(value) => value,
        Err(err) => {
            error!("{}: {:?}", message, err);
            process::exit(1);
        }
    }
}
