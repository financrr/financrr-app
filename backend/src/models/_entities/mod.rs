//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.10

pub mod prelude;

pub mod bank_accounts;
pub mod categories;
pub mod currencies;
pub mod external_bank_institutions;
pub mod file_attachments;
pub mod fixtures;
pub mod go_cardless_enduser_agreements;
pub mod go_cardless_requisitions;
pub mod imported_bank_accounts;
pub mod instances;
pub mod opensearch_migrations;
pub mod pending_transactions;
pub mod recurring_transactions;
pub mod sessions;
pub mod taggings;
pub mod tags;
pub mod transaction_parties;
pub mod transaction_templates;
pub mod transactions;
pub mod user_permissions;
pub mod users;
