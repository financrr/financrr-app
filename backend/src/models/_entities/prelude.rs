//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.1

pub use super::bank_accounts::Entity as BankAccounts;
pub use super::budget_criteria::Entity as BudgetCriteria;
pub use super::budget_criteria_bank_accounts::Entity as BudgetCriteriaBankAccounts;
pub use super::budget_criteria_categories::Entity as BudgetCriteriaCategories;
pub use super::budget_criteria_external_bank_accounts::Entity as BudgetCriteriaExternalBankAccounts;
pub use super::budget_criteria_tags::Entity as BudgetCriteriaTags;
pub use super::budget_histories::Entity as BudgetHistories;
pub use super::budgets::Entity as Budgets;
pub use super::categories::Entity as Categories;
pub use super::contracts::Entity as Contracts;
pub use super::currencies::Entity as Currencies;
pub use super::external_bank_account_ibans::Entity as ExternalBankAccountIbans;
pub use super::external_bank_accounts::Entity as ExternalBankAccounts;
pub use super::external_bank_institutions::Entity as ExternalBankInstitutions;
pub use super::file_attachments::Entity as FileAttachments;
pub use super::inactive_contracts::Entity as InactiveContracts;
pub use super::instances::Entity as Instances;
pub use super::linked_back_accounts::Entity as LinkedBackAccounts;
pub use super::pending_transactions::Entity as PendingTransactions;
pub use super::recurring_transactions::Entity as RecurringTransactions;
pub use super::sessions::Entity as Sessions;
pub use super::taggings::Entity as Taggings;
pub use super::tags::Entity as Tags;
pub use super::transaction_parties::Entity as TransactionParties;
pub use super::transaction_templates::Entity as TransactionTemplates;
pub use super::transactions::Entity as Transactions;
pub use super::user_permissions::Entity as UserPermissions;
pub use super::users::Entity as Users;
