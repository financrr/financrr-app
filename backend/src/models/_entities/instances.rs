//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.7

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "instances")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub node_id: i16,
    pub last_heartbeat: DateTimeWithTimeZone,
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}
