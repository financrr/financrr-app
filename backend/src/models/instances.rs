use super::_entities::instances::{ActiveModel, Column, Entity, Model};
use crate::services::instance_handler::INSTANCE_HEARTBEAT_TOLERANCE_SECONDS;
use loco_rs::model::ModelError;
use loco_rs::prelude::ModelResult;
use sea_orm::entity::prelude::*;
use sea_orm::{ActiveValue, IntoActiveModel, QueryOrder, QuerySelect, TransactionTrait};

pub type Instances = Entity;

#[async_trait::async_trait]
impl ActiveModelBehavior for ActiveModel {
    // extend activemodel below (keep comment for generators)

    async fn before_save<C>(self, _db: &C, insert: bool) -> std::result::Result<Self, DbErr>
    where
        C: ConnectionTrait,
    {
        if !insert && self.updated_at.is_unchanged() {
            let mut this = self;
            this.updated_at = sea_orm::ActiveValue::Set(chrono::Utc::now().into());
            Ok(this)
        } else {
            Ok(self)
        }
    }
}

impl Model {
    pub async fn get_node_id_and_create_new_instance(db: &DatabaseConnection) -> ModelResult<Self> {
        db.transaction::<_, _, ModelError>(|txn| {
            Box::pin(async move {
                let all_instances = Entity::find()
                    .order_by_asc(Column::NodeId)
                    .lock_exclusive() // Lock the rows to prevent concurrent access
                    .all(txn)
                    .await?;

                let all_node_ids = all_instances.iter().map(|model| model.node_id).collect::<Vec<_>>();

                let active_node_ids = all_instances
                    .iter()
                    .filter(|model| {
                        let last_heartbeat = model.last_heartbeat.to_utc();
                        let now = chrono::Utc::now();
                        let diff = now - last_heartbeat;
                        diff.num_seconds() < INSTANCE_HEARTBEAT_TOLERANCE_SECONDS as i64
                    })
                    .map(|model| model.node_id)
                    .collect::<Vec<_>>();

                let next_node_id = find_smallest_available_number(&all_node_ids, &active_node_ids);

                let model = Self::create_or_update_instance(txn, next_node_id).await?;

                Ok(model)
            })
        })
        .await
        .map_err(|err| ModelError::Any(err.into()))
    }

    pub async fn find_by_node_id(db: &impl ConnectionTrait, node_id: i16) -> ModelResult<Self> {
        let result = Entity::find().filter(Column::NodeId.eq(node_id)).one(db).await?;

        result.ok_or_else(|| ModelError::EntityNotFound)
    }

    pub async fn count_instances(db: &impl ConnectionTrait) -> ModelResult<u64> {
        Ok(Entity::find().count(db).await?)
    }

    pub async fn create_new_instance(db: &impl ConnectionTrait, node_id: i16) -> ModelResult<Self> {
        let instance = ActiveModel {
            node_id: ActiveValue::set(node_id),
            last_heartbeat: ActiveValue::set(chrono::Utc::now().into()),
            created_at: ActiveValue::set(chrono::Utc::now().into()),
            updated_at: ActiveValue::set(chrono::Utc::now().into()),
        };

        Ok(instance.insert(db).await?)
    }

    async fn create_or_update_instance(db: &impl ConnectionTrait, node_id: i16) -> ModelResult<Self> {
        let instance = Self::find_by_node_id(db, node_id).await;

        match instance {
            Ok(instance) => instance.into_active_model().update_heartbeat(db).await,
            Err(ModelError::EntityNotFound) => Self::create_new_instance(db, node_id).await,
            Err(err) => Err(err),
        }
    }
}

impl ActiveModel {
    pub async fn update_heartbeat(mut self, db: &impl ConnectionTrait) -> ModelResult<Model> {
        self.last_heartbeat = ActiveValue::Set(chrono::Utc::now().into());

        Ok(self.update(db).await?)
    }
}

fn find_smallest_available_number(all_node_ids: &[i16], active_node_ids: &[i16]) -> i16 {
    let active_set: std::collections::HashSet<i16> = active_node_ids.iter().copied().collect();

    for i in 0.. {
        if !active_set.contains(&(i as i16)) {
            return i as i16;
        }
    }

    all_node_ids.len() as i16
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_find_smallest_available_number() {
        assert_eq!(
            find_smallest_available_number(&[0, 1, 2, 4, 5, 6, 7], &[0, 1, 2, 4, 5, 6, 7]),
            3
        );
        assert_eq!(
            find_smallest_available_number(&[0, 1, 2, 3, 4, 5, 6, 7], &[0, 1, 2, 3, 4, 5, 6, 7]),
            8
        );
        assert_eq!(
            find_smallest_available_number(&[2, 3, 4, 5, 6, 7], &[2, 3, 4, 5, 6, 7]),
            0
        );
        assert_eq!(find_smallest_available_number(&[0, 1, 2, 3, 4, 5, 6, 7, 8, 9], &[9]), 0);
        assert_eq!(find_smallest_available_number(&[], &[]), 0);
    }
}
