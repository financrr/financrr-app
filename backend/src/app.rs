use crate::initializers::openapi::OpenApiInitializer;
use crate::initializers::services::ServicesInitializer;
use crate::utils::folder::{create_necessary_folders, STORAGE_FOLDER};
use crate::{controllers, models::_entities::users, tasks, workers::downloader::DownloadWorker};
use async_trait::async_trait;
use loco_rs::cache::Cache;
use loco_rs::storage::Storage;
use loco_rs::{
    app::{AppContext, Hooks, Initializer},
    bgworker::{BackgroundWorker, Queue},
    boot::{create_app, BootResult, StartMode},
    cache,
    controller::AppRoutes,
    db::{self, truncate_table},
    environment::Environment,
    storage,
    task::Tasks,
    Result,
};
use migration::Migrator;
use mimalloc::MiMalloc;
use sea_orm::DatabaseConnection;
use std::path::Path;

#[global_allocator]
static GLOBAL: MiMalloc = MiMalloc;

pub struct App;

#[async_trait]
impl Hooks for App {
    fn app_version() -> String {
        format!(
            "{} ({})",
            env!("CARGO_PKG_VERSION"),
            option_env!("BUILD_SHA").or(option_env!("GITHUB_SHA")).unwrap_or("dev")
        )
    }

    fn app_name() -> &'static str {
        env!("CARGO_CRATE_NAME")
    }

    async fn boot(mode: StartMode, environment: &Environment) -> Result<BootResult> {
        create_app::<Self, Migrator>(mode, environment).await
    }

    async fn initializers(_ctx: &AppContext) -> Result<Vec<Box<dyn Initializer>>> {
        Ok(vec![Box::new(OpenApiInitializer), Box::new(ServicesInitializer)])
    }

    fn routes(_ctx: &AppContext) -> AppRoutes {
        AppRoutes::with_default_routes() // controller routes below
            .prefix("/api/v1")
            .add_route(controllers::auth::routes())
            .add_route(controllers::session::routes())
            .add_route(controllers::openapi::routes())
    }

    async fn after_context(ctx: AppContext) -> Result<AppContext> {
        create_necessary_folders()?;

        Ok(AppContext {
            cache: Cache::new(cache::drivers::inmem::new()).into(),
            storage: Storage::single(storage::drivers::local::new_with_prefix(STORAGE_FOLDER)?).into(),
            ..ctx
        })
    }

    async fn connect_workers(ctx: &AppContext, queue: &Queue) -> Result<()> {
        queue.register(DownloadWorker::build(ctx)).await?;
        Ok(())
    }
    fn register_tasks(tasks: &mut Tasks) {
        tasks.register(tasks::seed::SeedData);
        // tasks-inject (do not remove)
    }

    async fn truncate(db: &DatabaseConnection) -> Result<()> {
        // TODO add all other tables
        truncate_table(db, users::Entity).await?;
        Ok(())
    }

    async fn seed(db: &DatabaseConnection, base: &Path) -> Result<()> {
        db::seed::<users::ActiveModel>(db, &base.join("users.yaml").display().to_string()).await?;
        Ok(())
    }
}
