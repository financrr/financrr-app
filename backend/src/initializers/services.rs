use crate::services::configure_services;
use async_trait::async_trait;
use axum::Router as AxumRouter;
use loco_rs::app::AppContext;
use loco_rs::prelude::Initializer;

pub struct ServicesInitializer;

#[async_trait]
impl Initializer for ServicesInitializer {
    fn name(&self) -> String {
        "Services".to_string()
    }

    async fn after_routes(&self, router: AxumRouter, ctx: &AppContext) -> loco_rs::Result<AxumRouter> {
        Ok(configure_services(router, ctx).await?)
    }
}
