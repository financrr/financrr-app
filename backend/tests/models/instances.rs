use crate::helpers::init::init_test;
use financrr::app::App;
use financrr::models::_entities::instances;
use financrr::services::Service;
use financrr::services::instance_handler::InstanceHandlerInner;
use financrr::services::snowflake_generator::MAX_NODE_ID;
use insta::assert_debug_snapshot;
use loco_rs::prelude::boot_test;
use serial_test::serial;

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("instances");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test(flavor = "multi_thread")]
#[serial]
async fn can_create_instance() {
    init_test!();

    let boot = boot_test::<App>().await.unwrap();

    let instance = instances::Model::get_node_id_and_create_new_instance(&boot.app_context.db).await;
    assert!(instance.is_ok());

    let instance = instance.unwrap();
    assert_eq!(instance.node_id, 1);
}

#[tokio::test(flavor = "multi_thread")]
#[serial]
async fn can_create_multiple_instances() {
    init_test!();

    let boot = boot_test::<App>().await.unwrap();

    let mut instances = Vec::new();

    for _ in 0..5 {
        let instance = instances::Model::get_node_id_and_create_new_instance(&boot.app_context.db).await;
        assert!(instance.is_ok());

        instances.push(instance.unwrap());
    }

    for (i, instance) in instances.into_iter().enumerate() {
        assert_eq!(instance.node_id, i as i16 + 1);
    }
}

#[tokio::test(flavor = "multi_thread")]
#[serial]
async fn can_create_instance_handler() {
    init_test!();

    let boot = boot_test::<App>().await.unwrap();

    let instance = InstanceHandlerInner::new(&boot.app_context).await;
    assert!(instance.is_ok());

    let node_id = instance.unwrap().get_instance_id();
    assert_eq!(node_id, 1);
}

#[tokio::test(flavor = "multi_thread")]
#[serial]
async fn can_create_maximum_instances() {
    init_test!();

    let boot = boot_test::<App>().await.unwrap();

    let mut instances = Vec::with_capacity(MAX_NODE_ID as usize);

    for _ in 0..MAX_NODE_ID {
        let instance = InstanceHandlerInner::new(&boot.app_context).await;
        assert!(instance.is_ok());

        instances.push(instance.unwrap());
    }

    for (i, instance) in instances.into_iter().enumerate() {
        assert_eq!(instance.get_instance_id(), i as u16 + 1);
    }

    let number_of_instances = instances::Model::count_instances(&boot.app_context.db).await.unwrap();
    assert_eq!(number_of_instances, MAX_NODE_ID + 1);

    // Assert that no new instance can be created
    let instance = InstanceHandlerInner::new(&boot.app_context).await;
    assert!(instance.is_err());

    let error = instance.err().unwrap();
    assert_debug_snapshot!(error)
}
