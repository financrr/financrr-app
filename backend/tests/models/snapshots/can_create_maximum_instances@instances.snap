---
source: tests/models/instances.rs
assertion_line: 97
expression: error
---
Any(
    NodeIdTooLarge,
)
